applications/convection_diffusion_application/custom_response_functions/local_temperature_average_response_function.h

```diff
@@ -0,0 +1,246 @@
+// KRATOS ___ ___  _  ___   __   ___ ___ ___ ___
+//       / __/ _ \| \| \ \ / /__|   \_ _| __| __|
+//      | (_| (_) | .` |\ V /___| |) | || _|| _|
+//       \___\___/|_|\_| \_/    |___/___|_| |_|  APPLICATION
+//
+//  License:       BSD License
+//                 Kratos default license: kratos/license.txt
+//
+//  Main authors:  Jordi Cotela
+//
+
+#ifndef KRATOS_LOCAL_TEMPERATURE_AVERAGE_RESPONSE_FUNCTION_H_INCLUDED
+#define KRATOS_LOCAL_TEMPERATURE_AVERAGE_RESPONSE_FUNCTION_H_INCLUDED
+
+#include "includes/kratos_flags.h"
+#include "includes/model_part.h"
+#include "response_functions/adjoint_response_function.h"
+
+namespace Kratos {
+///@addtogroup ConvectionDiffusionApplication
+///@{
+
+///@name Kratos Classes
+///@{
+
+class LocalTemperatureAverageResponseFunction: public AdjointResponseFunction
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(LocalTemperatureAverageResponseFunction);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+    LocalTemperatureAverageResponseFunction(Parameters Settings, ModelPart& rModelPart)
+    : mrModelPart(rModelPart)
+    , mNumNodes(0)
+    {
+        KRATOS_TRY;
+
+        std::string target_model_part = Settings["model_part_name"].GetString();
+
+        auto& r_target_model_part = GetTargetModelPart(rModelPart, target_model_part);
+        auto& r_nodes = r_target_model_part.Nodes();
+        mNumNodes = r_nodes.size();
+
+        //#pragma omp parallel for
+        for (int i = 0; i < mNumNodes; i++)
+        {
+            auto i_node = r_nodes.begin() + i;
+            i_node->Set(STRUCTURE);
+            i_node->SetValue(NUMBER_OF_NEIGHBOUR_ELEMENTS,0);
```

Comment:
VariableUtils::SetNonHistoricalVariableToZero
VariableUtils::SetFlag

Reply:
Agreed for VariableUtils::SetFlag but VariableUtils::SetNonHistoricalVariableToZero is not threadsafe if the variable is not initialized in advance, which is likely to be the case here.